File: ADSUS_BE.BLL/MedicalRecord/DTOs/CaseDTOs.cs
namespace ADSUS_BE.BLL.MedicalRecord.DTOs;

/// <summary>
/// #21, #22 và nhúng trong #23.
/// imageUrl là URL có hạn do Storage ký; NULL nếu ký thất bại (file không tồn tại).
/// KHÔNG BAO GIỜ trả fileRef — đó là đường dẫn lưu trữ thô.
/// </summary>
public sealed record UltrasoundImageResponse(
    Guid ImageId,
    Guid CaseId,
    string? ImageUrl,
    DateTime UploadedAt,
    string? Note);

/// <summary>1 loại thuốc trong đơn — nhúng trong PrescriptionSummary.</summary>
public sealed record PrescriptionItemSummary(
    string MedicineName,
    string Dosage,
    short DurationDays,
    DateOnly StartDate,
    string? Instructions);

/// <summary>
/// Đơn thuốc nhúng trong #23. Đính chính 15/08/2026: trước đây chỉ có PrescriptionId+Status
/// ("chi tiết thuộc Module 7") — Module 7 xác nhận CHƯA có màn nào xem chi tiết đơn thuốc, nên
/// giờ hiện đầy đủ thẳng ở đây (dùng chung cho cả CaseResponse lẫn PatientCaseResponse).
/// </summary>
public sealed record PrescriptionSummary(
    Guid PrescriptionId,
    string Status,
    DateOnly PrescribedDate,
    string? GeneralNote,
    IReadOnlyList<PrescriptionItemSummary> Items);

/// <summary>
/// #20, #23 — bản đầy đủ cho Bác sĩ/Điều dưỡng (Web SCR-12).
/// </summary>
public sealed record CaseResponse(
    Guid CaseId,
    Guid PatientProfileId,
    Guid DoctorId,
    string DoctorName,
    DateOnly VisitDate,
    string? ClinicalInfo,
    string Status,
    string? FinalDiagnosis,
    string? DoctorConclusion,
    PatientProfileResponse? PatientProfile,
    IReadOnlyList<UltrasoundImageResponse> UltrasoundImages,
    // AiResults removed
    PrescriptionSummary? Prescription,
    DateTime CreatedAt,
    DateTime UpdatedAt);

/// <summary>
/// #23 khi người gọi là Bệnh nhân (Mobile SCR-14).
///
/// Là một KIỂU RIÊNG chứ không phải CaseResponse với vài field để null: quy tắc dữ liệu nhạy
/// cảm nói field mà người gọi không có quyền thì không được KHAI BÁO, mà class C# thì không
/// ẩn field được — null vẫn serialize ra. Tách kiểu là cách duy nhất để trình biên dịch bảo
/// đảm clinicalInfo / patientProfile / aiResults không bao giờ tới tay bệnh nhân (GB-05).
///
/// Đính chính 15/08/2026: trước đây comment này còn liệt cả ultrasoundImages vào danh sách bị
/// cấm — SAI theo quyết định UCS 01/08/2026 (xem CasesController.cs's ExportReport doc comment
/// + design spec 2026-08-15): Patient CÓ xem được ảnh siêu âm gốc một khi ca đã Confirmed/End,
/// giống hệt nội dung PDF export (trừ chức năng xuất file). Chỉ dữ liệu nội bộ thuần Staff
/// (ClinicalInfo, PatientProfile, AiResults, các mốc CreatedAt/UpdatedAt) mới còn bị cấm.
/// </summary>
public sealed record PatientCaseResponse(
    Guid CaseId,
    Guid DoctorId,
    string DoctorName,
    DateOnly VisitDate,
    string Status,
    string? FinalDiagnosis,
    string? DoctorConclusion,
    PrescriptionSummary? Prescription,
    IReadOnlyList<UltrasoundImageResponse> UltrasoundImages);

/// <summary>#25 — một dòng trong danh sách lần khám CỦA CHÍNH bệnh nhân (Mobile).</summary>
public sealed record CaseSummaryResponse(
    Guid CaseId,
    DateOnly VisitDate,
    string Status,
    Guid DoctorId);

/// <summary>
/// #24 — một dòng trong danh sách lần khám cho Bác sĩ/Điều dưỡng (Web SCR-12).
///
/// Tách riêng khỏi CaseSummaryResponse (thêm 06/08/2026) chỉ để có thêm CreatedAt — VisitDate
/// là DateOnly, không có giờ. KHÔNG dùng chung với #25: PatientCaseResponse (chi tiết 1 ca cho
/// bệnh nhân) đã cố tình bỏ mọi mốc thời gian, nên danh sách của bệnh nhân cũng không nên có,
/// tránh lệch giữa hai màn cùng vai trò.
/// </summary>
public sealed record StaffCaseSummaryResponse(
    Guid CaseId,
    DateOnly VisitDate,
    string Status,
    Guid DoctorId,
    DateTime CreatedAt);

/// <summary>#20 — tạo lần khám mới kèm ít nhất một ảnh siêu âm (UC-07).</summary>
public sealed record CreateCaseRequest(
    Guid PatientProfileId,
    Guid ResponsibleDoctorId,
    string? ClinicalInfo,
    IReadOnlyList<UploadedFile> Images);

/// <summary>#21 — bổ sung ảnh vào một ca chưa được chốt (UC-07).</summary>
public sealed record AddUltrasoundImagesRequest(
    IReadOnlyList<UploadedFile> Images,
    string? Note);

/// <summary>
/// Thêm 07/08/2026, sửa lại cùng ngày (tách Lưu/Kết thúc) — Bác sĩ phụ trách nhập/sửa kết
/// luận cho ca khám ngay tại màn chi tiết ca (Module 04), không đợi màn duyệt kết quả AI
/// riêng (UC-19, đang được xây song song bởi một luồng công việc khác).
///
/// Dùng chung cho HAI hành động khác nhau, xem ICaseService:
///   SaveConclusionAsync — chỉ lưu nội dung, KHÔNG đổi trạng thái, sửa lại được nhiều lần.
///   ConfirmAsync        — lưu VÀ khoá ca (CONFIRMED), không có đường lùi (GB-01/P2).
/// </summary>
public sealed record CaseConclusionRequest(
    string FinalDiagnosis,
    string DoctorConclusion);

File: ADSUS_Mobile/lib/features/medical_record/data/dtos/case_dtos.dart
/// DTO khớp 1:1 `CaseSummaryResponse` — API Spec #25 (GET /cases/me).
class CaseSummaryDto {
  const CaseSummaryDto({
    required this.caseId,
    required this.visitDate,
    required this.status,
    required this.doctorId,
  });

  final String caseId;
  final String visitDate;
  final String status;
  final String doctorId;

  factory CaseSummaryDto.fromJson(Map<String, dynamic> json) => CaseSummaryDto(
        caseId: json['caseId'] as String,
        visitDate: json['visitDate'] as String,
        status: json['status'] as String,
        doctorId: json['doctorId'] as String,
      );
}

/// DTO khớp 1:1 `RelativeCaseSummaryResponse` — API (GET /cases/relatives).
/// Mở rộng CaseSummaryDto với tên bệnh nhân + quan hệ.
class RelativeCaseSummaryDto {
  const RelativeCaseSummaryDto({
    required this.caseId,
    required this.visitDate,
    required this.status,
    required this.doctorId,
    required this.patientName,
    this.relationshipName,
  });

  final String caseId;
  final String visitDate;
  final String status;
  final String doctorId;
  final String patientName;
  final String? relationshipName;

  factory RelativeCaseSummaryDto.fromJson(Map<String, dynamic> json) =>
      RelativeCaseSummaryDto(
        caseId: json['caseId'] as String,
        visitDate: json['visitDate'] as String,
        status: json['status'] as String,
        doctorId: json['doctorId'] as String,
        patientName: json['patientName'] as String? ?? 'Người thân',
        relationshipName: json['relationshipName'] as String?,
      );
}

/// DTO khớp 1:1 `PrescriptionItemSummary` lồng trong `PrescriptionSummaryDto`.
class PrescriptionItemDto {
  const PrescriptionItemDto({
    required this.medicineName,
    required this.dosage,
    required this.durationDays,
    required this.startDate,
    this.instructions,
  });

  final String medicineName;
  final String dosage;
  final int durationDays;
  final String startDate;
  final String? instructions;

  factory PrescriptionItemDto.fromJson(Map<String, dynamic> json) => PrescriptionItemDto(
        medicineName: json['medicineName'] as String,
        dosage: json['dosage'] as String,
        durationDays: json['durationDays'] as int,
        startDate: json['startDate'] as String,
        instructions: json['instructions'] as String?,
      );
}

/// DTO khớp 1:1 `PrescriptionSummary` lồng trong `CaseResponse`/`PatientCaseResponse` — API
/// Spec #23. Đính chính 15/08/2026: thêm prescribedDate/generalNote/items (backend đã trả sẵn,
/// Mobile trước đó không đọc — chỉ có prescriptionId/status).
class PrescriptionSummaryDto {
  const PrescriptionSummaryDto({
    required this.prescriptionId,
    required this.status,
    required this.prescribedDate,
    this.generalNote,
    this.items = const [],
  });

  final String prescriptionId;
  final String status;
  final String prescribedDate;
  final String? generalNote;
  final List<PrescriptionItemDto> items;

  factory PrescriptionSummaryDto.fromJson(Map<String, dynamic> json) =>
      PrescriptionSummaryDto(
        prescriptionId: json['prescriptionId'] as String,
        status: json['status'] as String,
        prescribedDate: json['prescribedDate'] as String,
        generalNote: json['generalNote'] as String?,
        items: (json['items'] as List<dynamic>? ?? const [])
            .map((e) => PrescriptionItemDto.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// DTO khớp 1:1 `UltrasoundImageResponse` lồng trong `CaseResponse`/`PatientCaseResponse`.
class UltrasoundImageDto {
  const UltrasoundImageDto({
    required this.imageId,
    required this.uploadedAt,
    this.imageUrl,
    this.note,
  });

  final String imageId;
  final String uploadedAt;
  final String? imageUrl;
  final String? note;

  factory UltrasoundImageDto.fromJson(Map<String, dynamic> json) => UltrasoundImageDto(
        imageId: json['imageId'] as String,
        uploadedAt: json['uploadedAt'] as String,
        imageUrl: json['imageUrl'] as String?,
        note: json['note'] as String?,
      );
}

/// DTO cho chẩn đoán bệnh theo danh mục chuẩn (CaseDiagnosisResponse).
class CaseDiagnosisDto {
  const CaseDiagnosisDto({
    required this.diagnosisItemId,
    required this.diagnosisName,
    required this.isOther,
    this.note,
  });

  final String diagnosisItemId;
  final String diagnosisName;
  final bool isOther;
  final String? note;

  factory CaseDiagnosisDto.fromJson(Map<String, dynamic> json) => CaseDiagnosisDto(
        diagnosisItemId: json['diagnosisItemId'] as String,
        diagnosisName: json['diagnosisName'] as String,
        isOther: json['isOther'] as bool,
        note: json['note'] as String?,
      );
}

/// DTO cho triệu chứng của ca khám (CaseSymptomResponse).
class CaseSymptomDto {
  const CaseSymptomDto({
    required this.categoryId,
    required this.categoryName,
    this.symptomId,
    this.symptomName,
    this.otherNote,
  });

  final String categoryId;
  final String categoryName;
  final String? symptomId;
  final String? symptomName;
  final String? otherNote;

  factory CaseSymptomDto.fromJson(Map<String, dynamic> json) => CaseSymptomDto(
        categoryId: json['categoryId'] as String,
        categoryName: json['categoryName'] as String? ?? '',
        symptomId: json['symptomId'] as String?,
        symptomName: json['symptomName'] as String?,
        otherNote: json['otherNote'] as String?,
      );
}

/// DTO khớp 1:1 `PatientCaseResponse` (field-set của Patient) — API Spec #23 (GET /cases/{id}).
///
/// Đính chính 15/08/2026: thêm doctorName/caseDiagnoses/ultrasoundImages (backend đã trả sẵn,
/// Mobile trước đó không đọc); đổi `conclusion` → `doctorConclusion` và SỬA key JSON đọc —
/// backend trả `doctorConclusion`, không phải `conclusion` (bug Critical, field trước đó luôn
/// null trong production dù backend có dữ liệu thật).
class CaseDto {
  const CaseDto({
    required this.caseId,
    required this.visitDate,
    required this.status,
    required this.doctorId,
    required this.doctorName,
    this.caseDiagnoses = const [],
    this.doctorConclusion,
    this.prescription,
    this.ultrasoundImages = const [],
    this.symptoms = const [],
  });

  final String caseId;
  final String visitDate;
  final String status;
  final String doctorId;
  final String doctorName;
  final List<CaseDiagnosisDto> caseDiagnoses;
  final String? doctorConclusion;
  final PrescriptionSummaryDto? prescription;
  final List<UltrasoundImageDto> ultrasoundImages;
  final List<CaseSymptomDto> symptoms;

  factory CaseDto.fromJson(Map<String, dynamic> json) => CaseDto(
        caseId: json['caseId'] as String,
        visitDate: json['visitDate'] as String,
        status: json['status'] as String,
        doctorId: json['doctorId'] as String,
        doctorName: json['doctorName'] as String,
        caseDiagnoses: (json['caseDiagnoses'] as List<dynamic>? ?? const [])
            .map((e) => CaseDiagnosisDto.fromJson(e as Map<String, dynamic>))
            .toList(),
        doctorConclusion: json['doctorConclusion'] as String?,
        prescription: json['prescription'] == null
            ? null
            : PrescriptionSummaryDto.fromJson(
                json['prescription'] as Map<String, dynamic>,
              ),
        ultrasoundImages: (json['ultrasoundImages'] as List<dynamic>? ?? const [])
            .map((e) => UltrasoundImageDto.fromJson(e as Map<String, dynamic>))
            .toList(),
        symptoms: (json['symptoms'] as List<dynamic>? ?? const [])
            .map((e) => CaseSymptomDto.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
